gpu/base/executor.cpp

```diff
@@ -46,34 +46,55 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 namespace gko {
 
 
-void CpuExecutor::raw_copy_to(const GpuExecutor *, size_type num_bytes,
+class device_guard {
+public:
+    device_guard(int device_id)
+    {
+        ASSERT_NO_CUDA_ERRORS(cudaGetDevice(&original_device_id));
+        ASSERT_NO_CUDA_ERRORS(cudaSetDevice(device_id));
+    }
+
+    ~device_guard()
```

Comment:
This is giving me the following warning:

```
/home/gflegar/Sources/ginkgo/public/gpu/base/executor.cpp:59:9: warning: '~device_guard' has a non-throwing exception specification but can still throw [-Wexceptions]
        ASSERT_NO_CUDA_ERRORS(cudaSetDevice(original_device_id));
        ^
/home/gflegar/Sources/ginkgo/public/core/base/exception_helpers.hpp:208:13: note: expanded from macro 'ASSERT_NO_CUDA_ERRORS'
            throw CUDA_ERROR(_errcode);   \
            ^
/home/gflegar/Sources/ginkgo/public/gpu/base/executor.cpp:57:5: note: destructor has a implicit non-throwing exception specification
    ~device_guard() // noexcept(false)
    ^
```

A simple "fix" would be to define it like this:

```c++
~device_guard() noexcept(false) { /* ... */ }
```

but I'm not sure if this is the behavior we want.

There are multiple approaches to fix this:

1.  we don't do anything about the warning - this will crash the program as soon as we raise an exception from within the destructor (I believe this is not the solution we want)
2.  we remove the `ASSERT` in the destructor and ignore the warnings - this may cause the library to just ignore some of the CUDA errors (this is what MAGMA often does, and I had a lot of problems because of it, as my program was crashing in some cases, and there was no way to fix it, even though I could have done something if MAGMA detected the warning and returned control to me), so I would recommend against doing this.
3.  we add  `noexcept(false)` to the definition - this will work in most cases but there is one scenario where the program crashes:
    1.  an exception is thrown in the scope where an object of `device_guard` type is allocated
    2.  the stack unwinding process begins and calls `~device_guard()`
    3.  assertion in `~device_guard()` fails due to an error in a previous asynchronous runtime API call, and raises another exception
    4.  program crashes as it is not possible to call exceptions while the stack unwinding process is in progress
4.  Instead of using the destructor to restore the originally active device, we instead define a special method `restore()` which does exactly the same thing as the destructor is doing now, and call it at the end of each function using `device_guard`. This won't cause any crashes, but will fail to restore the original device if an exception occurs within the operation / memory allocation / memory copy / synchronization.
5.  We use 3., but ignore the error code if stack unwinding is in progress (using [`std::uncaught_exception()`](http://en.cppreference.com/w/cpp/error/uncaught_exception)), this way we  never crash the user's program, and only ignore an error code if another error code is already being handled

@ginkgo-project/developers your thoughts?

Also note that there are some CUDA failures (like when using `assert()` or `asm("trap;");`) that will cause the runtime to continue returning the same error code until `cudaDeviceReset()` is called. If an operation causes this kind of error, then the scenario described in 3. is bound to happen. This is my motivation for suggesting 5. (it does the correct thing in this case).

Reply:
With what compiler/CUDA version is that? I weirdly did not see that so it looks like I was already in the scenario of 3. In which case I do not mind ignoring the error code on top of it. Maybe it's because I'm not very C++-ish, but when shit has hit the fan so much that what is described in 3. happens, I'm not really against letting the program crash...

Reply:
clang 5.0.1 and CUDA 9.1 (but I don't think that CUDA version matters much, as this is host code, which also has `.cpp` extension, so is compiled as C++ code).

Even if you didn't see the warning, your executable will still have the behavior of 1., as long as it's compiled with a C++11 or newer compiler. As part of the standard, all destructors are implicitly defined as `noexcept`, which means that weird stuff happens as soon as you throw an exception from the destructor.

> Maybe it's because I'm not very C++-ish, but when shit has hit the fan so much that what is described in 3. happens, I'm not really against letting the program crash...

If this wasn't CUDA runtime API, but anything else which has "normal" exception semantics, I would agree with you. However, CUDA has this weird concept of "everything is asynchronous, and every runtime API call can return a random error code from a previous call", and also has some weird error codes that are not cleared once they are returned.
Because of this it's very easy to trigger the behavior from 3.:

```c++
__global__ void k() { asm("trap;"); }

void f(const Executor *exec) { exec->run([](shared_ptr<const CpuExecutor>){}, [](shared_ptr<const GpuExecutor>) { k(); }); }
```
After calling `f` with a `GpuExecutor` as input, solution 3. will crash the program, while solution 5. will throw a `CudaException` with a description that there was a `trap` statement encountered in the kernel.

Reply:
I went with solution 3 + 5 for now because that looked the best to me. Still awaiting for feedback from other people though.